### ISSUE TITLE
Migrations are open again after upgrading from Neos 5.x to 7

### ISSUE THREAD

**Issue by daniellienert** [neos/flow-development-collection#2357]
While I had no open migrations before upgrade, it now shows 116 open migrations. 

<img width="1014" alt="Bildschirmfoto 2020-12-16 um 17 12 10" src="https://user-images.githubusercontent.com/642226/102374869-1f85cd80-3fc2-11eb-92b5-c766168f3ed2.png">

Flow/Neos 7

**Comment by daniellienert** [comment]
Ok, I read through the thread on slack and https://github.com/neos/flow-development-collection/issues/2244. 

Did I get it right, that I have to run `./flow doctrine:migrationstatus --show-migrations.`  before running `flow doctrine:migrate`. I fear that will braek a lot of automatic deployments.

**Comment by kdambekalns** [comment]
No, *whatever* you do with Doctrine migrations should fix the metadata storage, thee way it's handled in `doctrine/migrations`. And with the extra flush @bwaidelich added, no error whtsoever should block that…

Were there any errors during `doctrine:migrate`?

**Comment by daniellienert** [comment]
No errors in terms of exceptions. Migrate failed trying to apply an ancient typo3_.... migration.

**Comment by sorenmalling** [comment]
> Were there any errors during `doctrine:migrate`?

Maybe related, but i had a sync-metadata-storage error yesterday, when running `doctrine:migrationgenerate`

```
 soren@kevin  ~/Projects/dafis-distribution/application   development ●  ./flow doctrine:migrationgenerate
The metadata storage is not up to date, please run the sync-metadata-storage command to fix this issue.

  Type: Doctrine\Migrations\Exception\MetadataStorageError
  File: Packages/Libraries/doctrine/migrations/lib/Doctrine/Migrations/Exception/Me
        tadataStorageError.php
  Line: 13
```

**Comment by robertlemke** [comment]
for the record, @daniellienert, the same happened to me just now:

```
An exception occurred while executing 'ALTER TABLE typo3_typo3cr_domain_model_workspace DROP FOREIGN KEY typo3_typo3cr_domain_model_workspace_ibfk_1':

SQLSTATE[42000]: Syntax error or access violation: 1091 Can't DROP 'typo3_typo3cr_domain_model_workspace_ibfk_1'; check that column/key exists

The exception details have been logged to the Flow system log.
Exception in line 182 of /application/Packages/Libraries/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php: An exception occurred while executing 'ALTER TABLE typo3_typo3cr_domain_model_workspace DROP FOREIGN KEY typo3_typo3cr_domain_model_workspace_ibfk_1':

SQLSTATE[42000]: Syntax error or access violation: 1091 Can't DROP 'typo3_typo3cr_domain_model_workspace_ibfk_1'; check that column/key exists - See also: 20201218153113827fca.txt
```

**Comment by robertlemke** [comment]
I restored an older backup of ours site's database and tried again – this time the migration went through. I suspect that something went wrong during my first attempt (maybe a package wasn't refactored to the new migration classes and failed) and then the migration was stuck forever.

**Comment by sorenmalling** [comment]
> and then the migration was stuck forever

But that migraiton should never have been marked as "should run" in the first place. I don't expect that you were starting from scratch, so the failed executing was already executed?

**Comment by bwaidelich** [comment]
See comments on #2244 as well

**Comment by Benjamin-K** [comment]
I'm too getting a lot of open/new migrations:
<img width="1326" alt="Bildschirmfoto 2020-12-21 um 16 34 28" src="https://user-images.githubusercontent.com/3098031/102793629-79173f00-43aa-11eb-8141-a6c84bd217f1.png">

When i try to run migrations with `flow doctrine:migrate` i get the same error that @robertlemke got:
```
An exception occurred while executing 'ALTER TABLE typo3_typo3cr_domain_model_workspace DROP FOREIGN KEY typo3_typo3cr_domain_model_workspace_ibfk_1':

SQLSTATE[42000]: Syntax error or access violation: 1091 Can't DROP 'typo3_typo3cr_domain_model_workspace_ibfk_1'; check that column/key exists

The exception details have been logged to the Flow system log.
Exception in line 182 of /.../Packages/Libraries/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php: An exception occurred while executing 'ALTER TABLE typo3_typo3cr_domain_model_workspace DROP FOREIGN KEY typo3_typo3cr_domain_model_workspace_ibfk_1':

SQLSTATE[42000]: Syntax error or access violation: 1091 Can't DROP 'typo3_typo3cr_domain_model_workspace_ibfk_1'; check that column/key exists - See also: 20201221163600666dab.txt
```

Looks like flow tries to re-run old migrations (that are already applied). 

Any progress on this? Any way we can help with this?

**Comment by daniellienert** [comment]
@Benjamin-K it indeed looks like that this happens if your first doctrine:migrate hits a migration which haven't be migrated to the types method signatures. 
In that case the migration Table is left in a broken state.

Did that happen to you too?

Since when I originally opened that ticket, I migrated some more projects without further issues.

**Comment by sorenmalling** [comment]
@daniellienert 
> Since when I originally opened that ticket, I migrated some more projects without further issues.

Did you run any commands before your migration command?

**Comment by Benjamin-K** [comment]
@daniellienert 

> Did that happen to you too?

I don't know, if i'm in a broken state or not. How do i confirm this? 
I'm using MySQL and my migration table looks like this:

```SQL
CREATE TABLE `flow_doctrine_migrationstatus` (
  `version` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `executed_at` datetime DEFAULT NULL,
  `execution_time` int(11) DEFAULT NULL,
  PRIMARY KEY (`version`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

**Comment by kdambekalns** [comment]
@Benjamin-K That's the new format. You can revert to the old like @bwaidelich did in https://github.com/neos/flow-development-collection/pull/2333#issuecomment-742368449

The migration is done in [`ensureInitialized()`](https://github.com/doctrine/migrations/blob/3ba799f7a50fd2874783e792759d7167ae097425/lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php#L150) using [`needsUpdate()`](https://github.com/doctrine/migrations/blob/3ba799f7a50fd2874783e792759d7167ae097425/lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php#L169-L176) and [`updateMigratedVersionsFromV1orV2toV3()`](https://github.com/doctrine/migrations/blob/3ba799f7a50fd2874783e792759d7167ae097425/lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php#L217-L244)

Again, I see no reason why this would fail, now that https://github.com/neos/flow-development-collection/pull/2333 is in the code.

> Any way we can help with this?

Provide a DB dump, the exact same codebase and ideally MySQL version and settings to be able to reproduce this…

**Comment by kdambekalns** [comment]
> now that #2333 is in the code.

Doh! It isn't in the code, we never merged that! 😬

**Comment by Benjamin-K** [comment]
@kdambekalns I used an outdated Namespace for Doctrine Migration in Wegmeister.DatabaseStorage which i fixed already (See [here](https://github.com/die-wegmeister/Wegmeister.DatabaseStorage/compare/1.2.0...1.2.2)). Maybe that caused the error / broken db state. I will try to revert the migration table later on and retry to upgrade to Neos 7 and give feedback.

**Comment by Benjamin-K** [comment]
Ok, looks like reverting the migration table did the trick. I don't know why i extended a wrong AbstractMigration, but i think it was created by Neos, so maybe others will have the same problem too. So this could be sth. for the [Upgrade instructions](https://docs.neos.io/cms/references/upgrade-instructions/upgrade-instructions-5-3-7-0), isn't it?
I now could run all migrations without errors. There was 1 new migration. I hope that were all of them, but the system is running as expected.

**Comment by Benjamin-K** [comment]
OK, this was my fault. I forget to run `flow flow:core:migrate` on that one package, which should have fixed the problem from the beginning. Sry for that.

**Comment by bwaidelich** [comment]
@Benjamin-K Thanks for the update! The released version _was_ buggy since #2333 wasn't included (in `7.0.0` and `7.0.1`). Closing this one since with `7.0.2` this should be fixed now (I hope).
